src/lib.rs

```diff
@@ -0,0 +1,418 @@
+    prelude::*,
+    providers::{Http, Provider},
+};
+use eyre::Result;
+
+pub const INITAL_STATE_PATH: &str = "InitialState.csv";
+pub const HTTP_URL: &str = "https://eth.llamarpc.com";
+pub const ZK_SYNC_ADDR: &str = "0x32400084C286CF3E17e7B677ea9583e60a000324";
+pub const GENESIS_BLOCK: u64 = 16627460;
+pub const BLOCK_STEP: u64 = 128;
+
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("invalid Calldata: {0}")]
+    InvalidCalldata(String),
+
+    #[error("invalid StoredBlockInfo: {0}")]
+    InvalidStoredBlockInfo(String),
+
+    #[error("invalid CommitBlockInfo: {0}")]
+    InvalidCommitBlockInfo(String),
+
+    #[error("invalid compressed bytecode: {0}")]
+    InvalidCompressedByteCode(String),
+}
+
+pub fn create_initial_state() {
+    let _input = fs::read_to_string(INITAL_STATE_PATH).unwrap();
+    todo!();
+}
+
+pub async fn init_eth_adapter() -> (Provider<Http>, Contract) {
+    let provider =
+        Provider::<Http>::try_from(HTTP_URL).expect("could not instantiate HTTP Provider");
+
+    let abi_file = std::fs::File::open("./IZkSync.json").unwrap();
+    let contract = Contract::load(abi_file).unwrap();
+
+    (provider, contract)
+}
+
+fn parse_calldata(
+    l2_block_number: U256,
+    commit_blocks_fn: &Function,
+    calldata: Vec<u8>,
+) -> Result<Vec<CommitBlockInfoV1>> {
+    let mut parsed_input = commit_blocks_fn
+        .decode_input(&calldata[4..])
+        .map_err(|e| ParseError::InvalidCalldata(e.to_string()))?;
+
+    if parsed_input.len() != 2 {
+        return Err(ParseError::InvalidCalldata(format!(
+            "invalid number of parameters (got {}, expected 2) for commitBlocks function",
+            parsed_input.len()
+        ))
+        .into());
+    }
+
+    let new_blocks_data = parsed_input
+        .pop()
+        .ok_or_else(|| ParseError::InvalidCalldata("new blocks data".to_string()))?;
+    let stored_block_info = parsed_input
+        .pop()
+        .ok_or_else(|| ParseError::InvalidCalldata("stored block info".to_string()))?;
+
+    let abi::Token::Tuple(stored_block_info) = stored_block_info else {
+        return Err(ParseError::InvalidCalldata("invalid StoredBlockInfo".to_string()).into());
+    };
+
+    let abi::Token::Uint(previous_l2_block_number) = stored_block_info[0].clone() else {
+        return Err(ParseError::InvalidStoredBlockInfo(
+            "cannot parse previous L2 block number".to_string(),
+        )
+        .into());
+    };
+
+    let abi::Token::Uint(_previous_enumeration_index) = stored_block_info[2].clone() else {
+        return Err(ParseError::InvalidStoredBlockInfo(
+            "cannot parse previous enumeration index".to_string(),
+        )
+        .into());
+    };
+
+    //let previous_enumeration_index = previous_enumeration_index.0[0];
+    // TODO: What to do here?
+    // assert_eq!(previous_enumeration_index, tree.next_enumeration_index());
+
+    parse_commit_block_info(l2_block_number, &new_blocks_data)
+}
+
+fn parse_commit_block_info(
```

Comment:
Ditto: test for this too

Also, this function is almost 200 lines long, so let's split this up :smile:

Reply:
I think in cases like these it makes sense to abstract away the field extraction into some `struct` that can represent the extracted version of that `data` token that we pass in. So essentially that function would instantiate that `struct` using the data token, which the `struct` would parse internally allotting a field for each data field. Just an idea, feel free to play around with it :muscle:

Reply:
Yeah, that's no problem